core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -124,7 +127,9 @@ class TransactionCoordinator(brokerId: Int,
           val producerId = producerIdManager.generateProducerId()
           val createdMetadata = new TransactionMetadata(transactionalId = transactionalId,
             producerId = producerId,
+            lastProducerId = RecordBatch.NO_PRODUCER_ID,
```

Comment:
Just checking my understanding. If the expected producer id and epoch is non-empty and there is no current producerId defined, we have hit the "unknown producer" case.  I think we want to allow the request, but return the newly generated id. But I think actually we would hit the `isStaleProducerId` check below since the expected producer id does not match either the current or the (undefined) last producer id.

Reply:
Good catch, thanks. I've added a flag to `CoordinatorEpochAndTxnMetadata` to indicate the the `TransactionMetadata` was newly-created. which lets us distinguish the non-empty expected ID and epoch and no current producer id case.